File: core/datacap-spi/src/main/java/io/edurt/datacap/spi/executor/Executor.java
package io.edurt.datacap.spi.executor;

import org.slf4j.Logger;

public interface Executor
{
    default String name()
    {
        // Intercept the suffix Executor to mark the executor name
        return this.getClass().getSimpleName()
                .replace("Executor", "")
                .replace("Module", "");
    }

    default String description()
    {
        return String.format("Integrate %s executor", this.name());
    }

    /**
     * A description of the before function.
     *
     * @param configure The Pipeline object used for configuration.
     * @param logger The Logger object used for logging.
     * @return The result of the before function.
     */
    String before(Pipeline configure, Logger logger);

    PipelineResponse start(Pipeline configure);
}

File: executor/datacap-executor-spi/src/main/kotlin/io/edurt/datacap/executor/Executor.kt
package io.edurt.datacap.executor

import io.edurt.datacap.executor.configure.ExecutorRequest
import io.edurt.datacap.executor.configure.ExecutorResponse

interface Executor {
    fun name(): String {
        return this.javaClass
                .simpleName
                .removeSuffix("Executor")
    }

    fun start(request: ExecutorRequest): ExecutorResponse

    fun stop(request: ExecutorRequest): ExecutorResponse
}
